### PR TITLE
Audio: Module adapter: Add a new simple copy function

### PR DESCRIPTION
**Pull request by singalsu**
This patch adds a simplified version of simple copy for module adapter clients with one source and one sink. The overhead of discovering multiple source and sink buffers is avoided.

The module_adapter_bind() and module_adapter_unbind() functions are changed to detect multiple sources or sinks condition to control the used simple copy function.

### CODE REVIEW

File: src/audio/module_adapter/module_adapter.c [hunk 1]
```diff
@@ -825,6 +840,57 @@ static int module_adapter_simple_copy(struct comp_dev *dev)
...
+			goto out;
+		}
+
+		ret = 0;
+	}
+
+	/* consume from input buffer */
+	mod->total_data_consumed += mod->input_buffers[0].consumed;
+	if (mod->input_buffers[0].consumed)
+		audio_stream_consume(&source_c->stream, mod->input_buffers[0].consumed);
```
Comment (lyakh) [inline src/audio/module_adapter/module_adapter.c:877]: the original version uses `comp_update_buffer_{consume,produce}()` which also calls the notifier. Is this change on purpose?

File: src/audio/module_adapter/module_adapter.c [hunk 2]
```diff
@@ -1319,6 +1388,12 @@ int module_adapter_bind(struct comp_dev *dev, void *data)
 	/* set the source dev pointer */
 	mod_source_info->sources[source_index] = source_dev;
 
+	/* Disable 1 source 1 sink simple copy mode if multiple source buffers */
+	if (count_ready_or_higher_sources(mod_source_info) > 1) {
+		comp_info(dev, "module_adapter_bind(), multi_sink_or_source = true");
+		mod->multi_source_or_sink = true;
```
Comment (lyakh) [inline src/audio/module_adapter/module_adapter.c:1394]: this doesn't seem right. You shouldn't need to count those sources in certain states. You can do this perfectly deterministically on paths where that state is set or cleared. If `.bind()` adds a source in one of those states - then you know that the number has changed. If it adds one in a lower state - it doesn't change here, so no need to re-count. Same with `.unbind()` - do we allow unbinding ready sources?

File: src/audio/module_adapter/module_adapter.c [hunk 3]
```diff
@@ -825,6 +840,57 @@ static int module_adapter_simple_copy(struct comp_dev *dev)
 	return ret;
 }
 
+static inline int module_adapter_simple_copy_one_source_one_sink(struct processing_module *mod)
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c:843]: Not using same (dev -> mod) saves one operation. Could maybe change module_adapter_simple_copy() instead? Dev is simply mod->dev when needed.

File: src/audio/module_adapter/module_adapter.c [hunk 4]
```diff
@@ -710,7 +725,7 @@ module_single_source_setup(struct comp_dev *dev,
 	return num_output_buffers;
 }
 
-static int module_adapter_simple_copy(struct comp_dev *dev)
+static inline int module_adapter_simple_copy(struct comp_dev *dev)
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c:728]: Yes, I'm pretty sure I got same performance without, so the compiler in-lined. I thought making it explicit is better, but agree if that's the practise.

File: src/audio/module_adapter/module_adapter.c (hunk 1 above)
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c:877]: It's using part of Liam's patch. Copied accidentally, but this is better, right?

File: src/audio/module_adapter/module_adapter.c (hunk 2 above)
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c:1394]: I likely have a bug in unbind since there's fails in repeated suspend-resume test. Need to rethink this, thanks for commenting!

File: src/audio/module_adapter/module_adapter.c [hunk 5]
```diff
@@ -1319,6 +1405,12 @@ int module_adapter_bind(struct comp_dev *dev, void *data)
 	/* set the source dev pointer */
 	mod_source_info->sources[source_index] = source_dev;
 
+	/* Disable 1 source 1 sink simple copy mode if multiple source buffers */
+	if (module_adapter_multi_sink_source_check(dev)) {
+		comp_dbg(dev, "module_adapter_bind(), multi_sink_or_source = true");
+		mod->multi_source_or_sink = true;
```
Comment (ranj063) [inline src/audio/module_adapter/module_adapter.c:1411]: you can simplify this to mod->multi_source_or_sink = module_adapter_multi_sink_source_check(dev);

File: src/audio/module_adapter/module_adapter.c [hunk 6]
```diff
@@ -1353,6 +1445,12 @@ int module_adapter_unbind(struct comp_dev *dev, void *data)
 	if (source_index >= 0)
 		mod_source_info->sources[source_index] = NULL;
 
+	/* Re-enable 1 source 1 sink simple copy mode if single source buffer */
+	if (mod->simple_copy && !module_adapter_multi_sink_source_check(dev)) {
```
Comment (ranj063) [inline src/audio/module_adapter/module_adapter.c:1449]: do we need this simple_copy check here? if all we care about is whether a module has multi sources/sinks, it shouldnt matter right?

File: src/audio/module_adapter/module_adapter.c [hunk 7]
```diff
@@ -1275,6 +1342,25 @@ int module_adapter_get_attribute(struct comp_dev *dev, uint32_t type, void *valu
...
+	list_for_item(blist, &dev->bsource_list)
+		num_sources++;
+
+	list_for_item(blist, &dev->bsink_list)
+		num_sinks++;
+
+	comp_dbg(dev, "num_sources=%d num_sinks=%d", num_sources, num_sinks);
+	if (num_sources > 1 || num_sinks > 1)
+		return true;
+	else
```
Comment (ranj063) [inline src/audio/module_adapter/module_adapter.c:1360]: redundant else

File: src/audio/module_adapter/module_adapter.c (hunk 2 above)
Comment (ranj063) [reply src/audio/module_adapter/module_adapter.c:1411]: @singalsu please take this patch https://github.com/thesofproject/sof/pull/7805/commits/1e8b25299f2c9760c11abe4ffe6256c9ea36d99b and I think your IPC timeouts with unbind will go away

File: src/audio/module_adapter/module_adapter.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/module_adapter/module_adapter.c]: I'm fine either way as long as it's consistent :-) Also worth checking with other functions in the same file. As for saving an operation - not sure where? You need both in both anyway. So it's either
```
struct comp_dev *dev = mod->dev;
```
or
```
struct processing_module *mod = comp_get_drvdata(dev);
```
looks like both are just a single pointer dereference, but you don't know what the compiler makes of them, so, not easy to say with certainty

[the remark this answers is not in the document]
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c]: Seem to be the same, it's a macro despite that it looks like a function. Unfortunately trace still needs dev handle, all other stuff needs only the more narrow mod scope.

File: src/audio/module_adapter/module_adapter.c (hunk 2 above)
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c:1411]: Thanks, that solves it! I totally missed that scenario.

File: src/audio/module_adapter/module_adapter.c (hunk 6 above)
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c]: I removed this but, I wonder if it now causes an issue. If a module adapter client has not set simple copy the processing should never be handled by simple copy function. To speed up code, the module_adapter_copy() when calling module_adapter_simple_copy_1to1() is not checking mod->simple_copy.

I think I need to add to module_adapter_multi_sink_source_check() test for mod->simple_copy, and return true if simple_copy is false.

**Comment by ranj063** [comment]
SOFCI TEST

File: src/audio/module_adapter/module_adapter.c [hunk 8]
```diff
@@ -245,6 +245,20 @@ int module_adapter_prepare(struct comp_dev *dev)
...
+	if (mod->simple_copy && mod->num_input_buffers == 1 && mod->num_output_buffers == 1) {
+		mod->source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+		mod->sink = sink;
+		mod->multi_source_or_sink = false;
+	} else {
+		mod->source = NULL;
+		mod->sink = NULL;
+		mod->multi_source_or_sink = true;
+	}
+
```
Comment (btian1) [inline src/audio/module_adapter/module_adapter.c:261]: which part called first? prepare or bind? seems above code is duplicated with bind code.

Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c:261]: I don't think there's guaranteed order for those events. Do you see risk or ways to mitigate it?

**Comment by singalsu** [comment]
The saving with this rebased pipeline2.0 adapted patch is 3.4 MCPS from sum of 10s playback average numbers

```
                            Name,        Mtrace id, avg MCPS, max MCPS
            host-copier.0.playback,   comp:0 0x40000,     3.39,    11.42
                          gain.1.1,   comp:0 0x60000,     3.28,     9.11
                         mixin.1.1,   comp:0 0x20000,     3.27,     3.45
                        mixout.2.1,   comp:1 0x30000,     2.59,     3.31
                          gain.2.1,   comp:1 0x60001,     3.37,    16.81
    dai-copier.HDA.Analog.playback,   comp:1 0x40001,     4.04,     4.28
                                  ,                 ,    19.95,    48.39

                              Name,        Mtrace id, avg MCPS, max MCPS
            host-copier.0.playback,   comp:0 0x40000,     3.40,     6.89
                          gain.1.1,   comp:0 0x60000,     2.34,     8.22
                         mixin.1.1,   comp:0 0x20000,     2.54,     2.64
                        mixout.2.1,   comp:1 0x30000,     1.74,     2.33
                          gain.2.1,   comp:1 0x60001,     2.44,     8.48
    dai-copier.HDA.Analog.playback,   comp:1 0x40001,     4.13,     4.51
                                  ,                 ,    16.60,    33.06
```

File: src/include/sof/audio/module_adapter/module/generic.h [hunk 9]
```diff
@@ -184,6 +184,8 @@ struct processing_module {
 	struct output_stream_buffer *output_buffers;
 	uint32_t num_input_buffers; /**< number of input buffers */
 	uint32_t num_output_buffers; /**< number of output buffers */
+	struct comp_buffer *source; /**< single source buffer for multi_source_or_sink = false */
```
Comment (marcinszkudlinski) [inline src/include/sof/audio/module_adapter/module/generic.h:187]: Please don't call it sink/source - may be confusions with struct sof_source / struct sof_sink interface

`struct comp_buffer source_comp_buffer;` 

or simiar

File: src/include/sof/audio/module_adapter/module/generic.h [hunk 10]
```diff
@@ -203,6 +205,11 @@ struct processing_module {
 	 */
 	bool skip_src_buffer_invalidate;
 
+	/* False for module with one sink and one source to reduce module processing
+	 * overhead. True if multiple source or sink buffers.
+	 */
+	bool multi_source_or_sink;
```
Comment (marcinszkudlinski) [inline src/include/sof/audio/module_adapter/module/generic.h:211]: bool simple_1_1_copy ?
to be consistent with module_adapter_audio_stream_copy_1to1 that is called when true

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]
Rather minor - just propositions for new names, no pushing. Just please don't use sink/source :)

**Comment by singalsu** [comment]
The PR is updated according to @marcinszkudlinski 's proposals. The boolean is inverted with new name `mod->stream_copy_single_to_single` .

File: src/audio/module_adapter/module_adapter.c [hunk 11]
```diff
@@ -313,6 +313,21 @@ int module_adapter_prepare(struct comp_dev *dev)
...
+	 */
+	if (mod->num_input_buffers == 1 && mod->num_output_buffers == 1) {
+		mod->source_comp_buffer = list_first_item(&dev->bsource_list,
+							  struct comp_buffer, sink_list);
+		mod->sink_comp_buffer = sink;
+		mod->stream_copy_single_to_single = true;
+	} else {
+		mod->source_comp_buffer = NULL;
+		mod->sink_comp_buffer = NULL;
+		mod->stream_copy_single_to_single = false;
```
Comment (lyakh) [inline src/audio/module_adapter/module_adapter.c:328]: `.stream_copy_1to1`?

**Review by lyakh — COMMENTED** [review]
just a couple of very minor comments

File: src/audio/module_adapter/module_adapter.c (hunk 11 above)
Comment (singalsu) [reply src/audio/module_adapter/module_adapter.c:328]: I was thinking 1to1 looks too much like 1:1 that would give wrong impression about doing just memcpy() from buffer to buffer.

**Comment by singalsu** [comment]
Previous CI round 43 successful and 2 failing checks, failed Internal Intel CI System/merge/build with West, failed 
sof-ci/jenkins/pr-device-test/main-cavs2.5  with multiple suspend resume. New try...

**Review by btian1 — APPROVED** [review]
seems Ranjani trying other method with your PR for #7850 #7858 
This one should be fine for merge.

Just one concern, all our optimization is trying to figure out special case and use specific function to handle it, this have benefit for this special case, however, it does increase performance for worst case and code size, I am not clear why COE team  must get those module performance low for this, if there is no low power(38.4mhz）mode, it is not must for handle module with multiple routines.

**Comment by singalsu** [comment]
> seems Ranjani trying other method with your PR for #7850 #7858 This one should be fine for merge.
> 
> Just one concern, all our optimization is trying to figure out special case and use specific function to handle it, this have benefit for this special case, however, it does increase performance for worst case and code size, I am not clear why COE team must get those module performance low for this, if there is no low power(38.4mhz）mode, it is not must for handle module with multiple routines.

The previous version did not not enable the optimization, now it's using fix by @ranj063 to get the bind/unbind code to work correctly.

Yes, this is a best case optimization. But it's quite common that components have only one source and and one sink so I think this is worth it. It will be even better when pipeline 2.x let's components have linear shared buffers. Since this is small add there is not much code size overhead and it can be cleaned up when the shared buffers work is ready.

**Review by ranj063 — APPROVED** [review]

**Comment by singalsu** [comment]
42 successful and 3 failing checks
- Zephyr / build-linux (zmain, imx8 imx8x imx8m) (pull_request)
- sof-ci/jenkins/pr-device-test/main-ace --> verify-kernel-boot-log.sh, check-pause-resume-capture-100.sh
- sof-ci/jenkins/pr-device-test/main-cavs2.5 --> check-suspend-resume-5.sh

**Comment by singalsu** [comment]
SOFCI TEST

**Comment by singalsu** [comment]
Now 43 successful and 2 failing checks
- Zephyr / build-linux (zmain, imx8 imx8x imx8m) 
- sof-ci/jenkins/pr-device-test/main-cavs2.5 --> check-suspend-resume-with-playback.sh timeout

**Comment by singalsu** [comment]
SOFCI TEST

**Comment by ranj063** [comment]
SOFCI TEST

**Comment by singalsu** [comment]
SOFCI TEST

**Review by kv2019i — APPROVED** [review]
Thanks @singalsu , looks good now!

**Comment by kv2019i** [comment]
https://sof-ci.01.org/sofpr/PR7792/build10174/devicetest/index.html looks a bit worrisome. I went though all the failures (verify-kernel-bootlog, check-runtime-pm-status , multiple-pipeline-capture ) and all are failures of https://github.com/thesofproject/sof/issues/7482 , IOW with MOD_SET_DX failing. I'm going to proceed and merge this, but we need to raise attention to #7482 .
